### PR TITLE
Memory bindings

### PR DESCRIPTION
**Pull request by michael-weigelt**
This adds bindings for Memory-associated functions. Memory can be imported, exported, read and manipulated.

### CODE REVIEW

File: lib/Wasmtime.hs [hunk 1]
```diff
@@ -952,12 +980,169 @@ withExterns externs f = allocaArray n $ \externs_ptr0 ->
...
+--------------------------------------------------------------------------------
+
+-- | A descriptor for a WebAssembly memory type.
+--
+-- Memories are described in units of pages (64KB) and represent contiguous chunks of addressable memory.
+newtype MemoryType = MemoryType {unMemoryType :: ForeignPtr C'wasm_memorytype_t}
+
+-- | Creates a descriptor for a WebAssembly 'Memory' with the specified minimum number of memory pages,
+-- an optional maximum of memory pages, and a 64 bit flag, where false defaults to 32 bit memory.
+newMemoryType :: Word64 -> Maybe Word64 -> Bool -> MemoryType
```
Comment (basvandijk) [inline lib/Wasmtime.hs:1000]:
```suggestion
newMemoryType :: 
  -- | Minimum number of memory pages.
  Word64 -> 
  -- | Optional maximum of memory pages.
  Maybe Word64 -> 
   -- | 'True' means memory is 64 bit flag and 'False' means 32 bit.
  Bool -> 
  MemoryType
```

File: test/memory_import.wat [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+  (func (export "size") (result i32) (memory.size))
+  (func (export "load") (param i32) (result i32)
+    (i32.load8_s (local.get 0))
+  )
+  (func (export "store") (param i32 i32)
+    (i32.store8 (local.get 0) (local.get 1))
+  )
+
+  (data (i32.const 0x1000) "\01\02\03\04")
+)
```
Comment (basvandijk) [inline test/memory_import.wat:13]: Is this file still necessary?

**Review by basvandijk — APPROVED** [review]

File: lib/Wasmtime.hs [hunk 3]
```diff
@@ -952,12 +980,169 @@ withExterns externs f = allocaArray n $ \externs_ptr0 ->
...
+readMemoryAt :: MonadPrim s m => Context s -> Memory s -> Word64 -> Word64 -> m (Either MemoryAccessError B.ByteString)
+readMemoryAt ctx mem offset len = do
+  max_len <- getMemorySizeBytes ctx mem
+  unsafeIOToPrim $ do
+    if offset + len > max_len
+      then pure $ Left MemoryAccessError
+      else do
+        res <- unsafeWithMemory ctx mem $ \mem_data_ptr mem_size ->
+          BI.create mem_size $ \dst_ptr ->
+            BI.memcpy dst_ptr mem_data_ptr mem_size
```
[the remark this answers is not in the document]
Comment (michael-weigelt) [reply lib/Wasmtime.hs:1118]: brilliant. copy-pasted and then only corrected the original. thanks for catching it!

File: test/memory_import.wat (hunk 2 above)
Comment (michael-weigelt) [reply test/memory_import.wat:13]: it is not in a c-inspired test. but I want to use it in a further test. 
should we make those live in a new directory?
